Gradle modify tool

ericssy opened the pull request:
I'm working with @byjiang1996 on making iDFlakies compatible with Gradle. This PR is created in reference to #30. It contains the Modify Gradle tool and a README file.

winglam commented on README-gradle.md:
```diff
@@ -0,0 +1,97 @@
+         ...
+         classpath group: 'edu.illinois.cs',
+         name: 'idflakies', 
+         version: '1.2.0'
```
I have deployed iDFlakies as version 1.1.0 instead (the version should be live in about one day). Please update this document globally

winglam commented on README-gradle.md:
```diff
@@ -0,0 +1,97 @@
+
+
+
+
```
Please remove these empty lines

winglam commented on gradle_modify/modify_gradle.sh:
```diff
@@ -0,0 +1,30 @@
+#!/bin/bash
+
+ARTIFACT_ID="idflakies"
+ARTIFACT_VERSION="1.0.3-SNAPSHOT"
```
Please use the versions matching what I changed in [this commit](https://github.com/idflakies/iDFlakies/commit/4bb34655975f687ea80c50af6e74cab760a21ad0)

winglam commented on gradle_modify/modify_gradle.sh:
```diff
@@ -0,0 +1,30 @@
+
+cd ${working_dir}
+
+# find ${project_path} -name build.gradle | grep -v "src/" | python modify_gradle.py ${ARTIFACT_ID} ${ARTIFACT_VERSION} ${CONFIGURATION_CLASS}
```
Please remove this line if it is not needed

winglam commented on README-gradle.md:
```diff
@@ -0,0 +1,97 @@
+## Using iDFlakies on a Gradle project
```
Please clarify what Gradle/Scala/Java/... versions are needed to use the Gradle plugin here

winglam replied:
Please move the dependency requirements to be first

winglam commented:
Since iDFlakies version 1.1.0 is now live, please double-check that the README instructions and modify Gradle/maven scripts work with the newly released version https://mvnrepository.com/artifact/edu.illinois.cs/idflakies/1.1.0

ericssy commented:
Hi Wing, 

Thank you for the feedback! We just discovered some issues of the gradle modify tool on certain projects. I'll push the commits once we are done with the testing.

ericssy commented:
Hi, Wing. I just updated the version and the README-gradle.md according to your comments. I also updated the strategy in modifying  Gradle projects with subprojects.

winglam commented on README-gradle.md:
```diff
@@ -0,0 +1,98 @@
+## Using iDFlakies on a Gradle project
+
+### Install iDFlakies and testrunner 
```
iDFlakies and testrunner do not need to installed.

The version of testrunner (1.2) and iDFlakies (1.1.0) containing the changes to support Gradle has already been released on Maven central. Please just have your script automatically set those versions in the build.gradle

You may also have instructions to use SNAPSHOT versions that would require you to build the two tools, but please clearly label such instructions and put them towards the bottom

winglam commented on README-gradle.md:
```diff
@@ -0,0 +1,98 @@
+
+### Automatically setting up the build.gradle for iDFlakies
+
+Run the following command to  automatically setup the build.gradle for iDFlakies.
```
Extra space

winglam replied:
This is still not addressed

winglam commented on README-gradle.md:
```diff
@@ -0,0 +1,98 @@
+
+2. Then, add `apply plugin: 'testrunner'` to the build.gradle file. The location of this code snippet depends on whether build.gradle file has `allprojects{}` and `subprojects{}`. 
+
+   * If there are both `allprojects{}` and `subprojects{}` methods in gradle.build, then the `apply plugin: "testrunner"` command should be added to the end of  `subprojects{}` and the root (outside of all methods in the build file).
```
Extra space

winglam commented on README-gradle.md:
```diff
@@ -0,0 +1,98 @@
+2. Then, add `apply plugin: 'testrunner'` to the build.gradle file. The location of this code snippet depends on whether build.gradle file has `allprojects{}` and `subprojects{}`. 
+
+   * If there are both `allprojects{}` and `subprojects{}` methods in gradle.build, then the `apply plugin: "testrunner"` command should be added to the end of  `subprojects{}` and the root (outside of all methods in the build file).
+   * If only `allprojects{}`method is in gradle.build, the command is added to  `allprojects{}` only
```
Missing period

winglam commented on README-gradle.md:
```diff
@@ -0,0 +1,98 @@
+   * If only `subprojects{}` method is in gradle.build, the command is added to both the end of `subprojects{}` and the root. 
+   * If there's no `allprojects{}` and `subprojects{}` in gradle.build, only add the command to the root.
+
+   Here is an example where this code snippet is added outside of all methods.
```
Please just provide an example for all 4 cases

winglam commented on README-gradle.md:
````diff
@@ -0,0 +1,98 @@
+   
+   apply plugin: 'testrunner'
+   ```
+3. For each subproject with build.gradle file in its directory, append `apply plugin: 'testrunner'` to the end of the build.gradle in the subproject. 
````
(1) Just to be clear, all subprojects with build.gradle files will need to be modified as well right?

(2) If (1) is true, then does the script automatically modify all subprojects' build.gradle files?

ericssy replied:
(1) Yes, all subprojects with build.gradle will need to be modified. 
(2) Yes, the script would automatically modify all of them.

winglam reviewed (commented):
Thanks a lot for the changes. Please look into the following the comments

ericssy commented:
Hi Wing, I just updated the README-gradle.md. Here are my chages: 
* Deleted extra spaces and lines 
* Added examples to all 4 cases 
* Moved dependency requirement to the top 
* Added instruction on using locally built iDFlakies

winglam commented on README-gradle.md:
```diff
@@ -0,0 +1,191 @@
+       
+       allprojects {
+         repositories {
+         	mavenCentral()
```
Please check the indentation of the code samples. ```mavenCentral()``` here seem to be indented more than line 44

winglam commented on README-gradle.md:
````diff
@@ -0,0 +1,191 @@
+```
+mvn clean install
+```
+```groovy
````
Please add some text before this code snippet. I'm assuming the point here is that if one chooses to use locally built version of iDFlakies they would have to modify the build.gradle similar to what is shown in this code snippet right? 

You may also consider adding the command to automatically use the ```gradle_modify/modify_gradle.sh``` script

winglam commented on README-gradle.md:
````diff
@@ -0,0 +1,191 @@
+}
+```
+### Running iDFlakies on Gradle project 
+After the build.gralde has been configured, one can run iDFlakies on the Gradle project with the following command: 
````
Mispelled word `gralde`

winglam reviewed (commented):
Please look to address the minor comments in this review. Overall the changes look good. Can you please specify a few projects in which you tested the modify_gradle.sh on?

ericssy commented:
Hi Wing, I just pushed some changes according to your feedback.

Please look at the most recent commit f1fecf8ad3ea57a0d8469781183650ba7213bcc2, as some of the changes are not reflected on previous ones. 

Changes: 
  * Fixed incorrect indentation 
  * Fixed typo
  * Added more instructions on how to use locally built iDFlakies 
  * Added the link to a spreadsheet with all the projects tested

winglam commented:
LGTM. Thanks again for all of the changes!
